src/modules/linux-micro/kmod/kmod.c

```diff
@@ -0,0 +1,273 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <libkmod.h>
+#include <stdio.h>
+
+#define SOL_LOG_DOMAIN &_log_domain
+#include "sol-log-internal.h"
+SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "linux-micro-kmod");
+
+#include "sol-platform-linux-micro.h"
+#include "sol-util.h"
+#include "sol-vector.h"
+
+static int
+kmod_apply_value(struct kmod_ctx *ctx, char *mod_name)
+{
+    const int probe_flags = KMOD_PROBE_APPLY_BLACKLIST;
+    struct kmod_list *itr, *modlist = NULL;
+    int r = 0;
+
+    SOL_DBG("load: %s", mod_name);
+
+    r = kmod_module_new_from_lookup(ctx, mod_name, &modlist);
+    if (r < 0) {
+        SOL_ERR("Failed to lookup alias '%s': %m", mod_name);
+        return r;
+    }
+
+    if (!modlist) {
+        SOL_ERR("Failed to find module '%s'", mod_name);
+        return -ENOENT;
+    }
+
+    kmod_list_foreach(itr, modlist) {
+        struct kmod_module *mod;
+        int state, err;
+
+        mod = kmod_module_get_module(itr);
+        state = kmod_module_get_initstate(mod);
+
+        switch (state) {
+        case KMOD_MODULE_BUILTIN:
+            SOL_INF("Module '%s' is builtin", kmod_module_get_name(mod));
+            break;
+
+        case KMOD_MODULE_LIVE:
+            SOL_DBG("Module '%s' is already loaded", kmod_module_get_name(mod));
+            break;
+
+        default:
+            err = kmod_module_probe_insert_module(mod, probe_flags,
+                NULL, NULL, NULL, NULL);
+
+            if (err == 0)
+                SOL_INF("Inserted module '%s'", kmod_module_get_name(mod));
+            else if (err == KMOD_PROBE_APPLY_BLACKLIST)
+                SOL_INF("Module '%s' is blacklisted", kmod_module_get_name(mod));
+            else {
+                SOL_ERR("Failed to insert '%s': %m", kmod_module_get_name(mod));
+                r = err;
+            }
+        }
+
+        kmod_module_unref(mod);
+    }
+
+    kmod_module_unref_list(modlist);
+    return r;
+}
+
+static char *
+remove_leading_whitespace(char *input)
+{
+    while (isspace(*input))
+        input++;
+    return input;
+}
+
+static void
+remove_trailing_whitespace(char *input)
+{
+    char *ptr = input + strlen(input) - 1;
+
+    while (ptr > input && isspace(*ptr)) {
+        *ptr = '\0';
+        ptr--;
+    }
+}
+
+static int
+kmod_apply_file(struct kmod_ctx *ctx, int fd)
+{
+    FILE *conf;
+    char buf[1024];
+    int err = 0, line_no = 1;
+
+    conf = fdopen(fd, "re");
+    if (!conf) {
+        close(fd);
+        return -errno;
+    }
+
+    while (fgets(buf, sizeof(buf), conf)) {
```

Comment:
I'd avoid doing `fgets()` as lines bigger than `sizeof(buf)` will screw you. And it's slower, then use the `sol_file_reader` to get a slice and operate on that. If possible it will `mmap()`.